Add rendered preview in admin. Fixes: #325

WoLpH opened the pull request.

jrief commented on post_office/admin.py:
```diff
@@ -1,3 +1,6 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
```
please remove the two lines above. In python-3 they don't make any sense.

WoLpH replied:
I'm not sure where those are coming from to be honest. I've never added them so it might be related to git merging things

jrief commented on post_office/admin.py:
```diff
@@ -9,6 +12,7 @@
 from django.core.mail.message import SafeMIMEText
 from django.forms import BaseInlineFormSet
 from django.forms.widgets import TextInput
+from django.utils import html
```
check order of import statements

jrief commented on post_office/admin.py:
```diff
@@ -256,11 +262,33 @@ class EmailTemplateInline(admin.StackedInline):
+    def rendered_content(self, instance):
+        if instance.content:
+            height = instance.content.count('\n') * 25
+            return html.mark_safe(
```
use `format_html`

jrief commented on post_office/admin.py:
```diff
@@ -256,11 +262,33 @@ class EmailTemplateInline(admin.StackedInline):
+
+    def rendered_html_content(self, instance):
+        if instance.html_content:
+            return html.mark_safe('<iframe '
```
detto

jrief commented on post_office/admin.py:
```diff
@@ -270,18 +298,69 @@ class EmailTemplateAdmin(admin.ModelAdmin):
     form = EmailTemplateAdminForm
     list_display = ('name', 'description_shortened', 'subject', 'languages_compact', 'created')
     search_fields = ('name', 'description', 'subject')
+    readonly_fields = 'rendered_content', 'rendered_html_content'
```
no implicit tuples

jrief commented on post_office/admin.py:
```diff
@@ -270,18 +298,69 @@ class EmailTemplateAdmin(admin.ModelAdmin):
+
+            return http.HttpResponse(template.render(instance.example_context))
+
+        return super(EmailTemplateAdmin, self).change_view(
```
this is Python-2 `super()`-style.

jrief commented on post_office/admin.py:
```diff
@@ -270,18 +298,69 @@ class EmailTemplateAdmin(admin.ModelAdmin):
+    def rendered_content(self, instance):
+        if instance.content:
+            height = instance.content.count('\n') * 25
+            return html.mark_safe(
```
detto

jrief commented on post_office/admin.py:
```diff
@@ -270,18 +298,69 @@ class EmailTemplateAdmin(admin.ModelAdmin):
+
+    def rendered_html_content(self, instance):
+        if instance.html_content:
+            return html.mark_safe('<iframe '
```
detto

jrief reviewed (changes requested):
Before fixing my code annotations, I'd rather rethink, if it doesn't make more sense to se a WYSIWYG-Editor straight ahead. This would be my preferred long term solution.

Thanks for your work, but this new feature must be approved by @selwin anyway. As I said, this shall only be a temporary solution.

WoLpH commented:
The pull request definitely needs cleanup, but since I didn't get a reply in the original issue anymore I thought it would be good to create a pull request so it doesn't get lost :)

Personally I think that this addition is useful even if you would have a WYSIWYG editor. The renderings of those editors are almost always different from a regular render. It's not my call however, if the WYSIWYG editor is the preferred solution than I'll close the pull request and wait for that release.

With that in mind... a form to send out a test-mail would be very useful as well.

selwin commented:
Having rendered HTML preview in admin would be great.

Aside from @jrief 's comments, please make sure that all JS are stripped to prevent XSS attacks.

WoLpH commented:
> Having rendered HTML preview in admin would be great.
> 
> Aside from @jrief 's comments, please make sure that all JS are stripped to prevent XSS attacks.

The easiest and safest way would be to sandbox the iframe, is that enough or should I also do a few search/replaces?
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox

selwin commented:
@WoLpH we have https://bleach.readthedocs.io/ as optional dependency. We can use that to easily strip all JS.

selwin commented:
I think we can even make `bleach` a required dependency @jrief . This rendered preview is a big quality of life improvement in my opinion.

WoLpH commented:
I've cleaned up the commits and added the `clean_html()` call.

I should note that there is another similar attack vector in the code-base outside of this addition:
https://github.com/ui/django-post_office/blob/5297f00d886406de1233de6697262096e9ea81ca/post_office/admin.py#L179-L182

And possibly this one:
https://github.com/ui/django-post_office/blob/5297f00d886406de1233de6697262096e9ea81ca/post_office/admin.py#L173-L177

selwin commented on post_office/models.py:
```diff
@@ -269,6 +269,7 @@ class EmailTemplate(models.Model):
         default='', blank=True)
     default_template = models.ForeignKey('self', related_name='translated_templates',
         null=True, default=None, verbose_name=_('Default template'), on_delete=models.CASCADE)
+    example_context = context_field_class(_('Context'), blank=True, null=True)
```
Why are we adding a new field to the model? Adding a preview to admin shouldn't necessitate adding a new field.

WoLpH replied:
I think the variable name is self explanatory, but it's to add a context to the rendered preview so all variables can be rendered with a useful value.

selwin commented on post_office/admin.py:
```diff
@@ -249,36 +256,111 @@ def __init__(self, *args, **kwargs):
     }
+    change_form_template = 'admin/post_office/EmailTemplate/change_form.html'
+
+    def change_view(self, request, object_id, form_url='', extra_context=None):
```
Can you move this dedicated preview functionality to a new view?

selwin commented:
@WoLpH gentle reminder to update this PR so we can merge this in.

gabn88 commented:
@WoLpH @selwin This sounds like a great feature that I might use right away in my project. I have been reading the code and the idea from the beginning. And it makes me ask myself, should we need an extra STATUS_CHOICES on the Email Model? Something like 'draft'? Or am I missing something?
